File: src/Pcl.NET/VectorInt.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Runtime.CompilerServices;
using System.Text;
using System.Threading.Tasks;

namespace Pcl.NET
{
    public class VectorInt : Vector<int>
    {
        public override IntPtr Data
        {
            get
            {
                return Invoke.std_vector_int_data(_ptr);
            }
        }
        public override long Count
        {
            get
            {
                ThrowIfDisposed();
                return (long)Invoke.std_vector_int_size(_ptr);
            }
        }

        public VectorInt()
        {
            _ptr = Invoke.std_vector_int_ctor();
        }

        public VectorInt(long count)
        {
            _ptr = Invoke.std_vector_int_ctor_count((ulong)count);
        }
        internal VectorInt(IntPtr ptr)
        {
            _ptr = ptr;
            _suppressDispose = true;
        }
        public unsafe VectorInt(int[] array)
        {
            _ptr = Invoke.std_vector_int_ctor_count((ulong)array.Length);
            CopyFromArray(array, DataU);
        }
        public override void Add(int item)
        {
            ThrowIfDisposed();
            Invoke.std_vector_int_add(_ptr, item);
        }

        public override void At(long index, ref int value)
        {
            ThrowIfDisposed();
            Invoke.std_vector_int_at(_ptr, (ulong)index, ref value);
        }

        public override void Clear()
        {
            ThrowIfDisposed();
            Invoke.std_vector_int_clear(_ptr);
        }

        public override void Insert(long index, int item)
        {
            ThrowIfDisposed();
            Invoke.std_vector_int_insert(_ptr, (ulong)index, item);
        }

        public override void Resize(long size)
        {
            ThrowIfDisposed();
            Invoke.std_vector_int_resize(_ptr, (ulong)size);
        }

        public override IEnumerator<int> GetEnumerator()
        {
            long count = Count;
            for (var i = 0; i < count; i++)
            {
                yield return this[i];
            }
        }

        protected override void DisposeObject()
        {
            if (!_suppressDispose)
            {
                Invoke.std_vector_int_delete(ref _ptr);
            }
        }

    }
}

File: src/pcl-bindings/VectorInt.cpp
#include "export.h"
#include <vector>

using point_t = int;
using vector_t = std::vector<int>;

EXPORT(vector_t*) std_vector_int_ctor()
{
    return new vector_t();
}

EXPORT(vector_t*) std_vector_int_ctor_count(size_t count)
{
    return new vector_t(count);
}

EXPORT(void) std_vector_int_delete(vector_t** ptr)
{
    delete* ptr;
    *ptr = NULL;
}

EXPORT(void) std_vector_int_at(vector_t* ptr, size_t idx, point_t* value)
{
    *value = ptr->at(idx);
}

EXPORT(size_t) std_vector_int_size(vector_t* ptr)
{
    return ptr->size();
}

EXPORT(void) std_vector_int_clear(vector_t* ptr)
{
    ptr->clear();
}

EXPORT(void) std_vector_int_resize(vector_t* ptr, size_t size)
{
    ptr->resize(size);
}

EXPORT(void) std_vector_int_add(vector_t* ptr, point_t value)
{
    ptr->push_back(value);
}

EXPORT(void) std_vector_int_insert(vector_t* ptr, ptrdiff_t index, point_t value)
{
    ptr->insert(ptr->begin() + index, value);
}

EXPORT(int*) std_vector_int_data(vector_t* ptr)
{
    return ptr->data();
}

